File: fix_pnpm.py
import glob
import re

workflows = glob.glob('.github/workflows/**/*.yml', recursive=True)

for wf in workflows:
    with open(wf) as f:
        content = f.read()

    # The previous script might have accidentally removed `version: ...` that wasn't `9.12.0` because we had some buggy regexes
    # We should only touch exactly `version: 9.12.0` (or similar quotes) inside a `with:` block under `uses: pnpm/action-setup@v[0-9]`

    # We already did `git reset --hard HEAD~1` so we are back to the original files.
    # Let's just do a simple string replace for the ones we know are bad.

    # First let's just find and replace the EXACT block that's causing issues.
    content = content.replace("uses: pnpm/action-setup@v4\n        with:\n          version: 9.12.0\n", "uses: pnpm/action-setup@v4\n")
    content = content.replace("uses: pnpm/action-setup@v3\n        with:\n          version: 9.12.0\n", "uses: pnpm/action-setup@v3\n")
    content = content.replace("uses: pnpm/action-setup@v2\n        with:\n          version: 9.12.0\n", "uses: pnpm/action-setup@v2\n")
    content = content.replace("uses: pnpm/action-setup@v4\n        with:\n          version: '9.12.0'\n", "uses: pnpm/action-setup@v4\n")
    content = content.replace("uses: pnpm/action-setup@v3\n        with:\n          version: '9.12.0'\n", "uses: pnpm/action-setup@v3\n")
    content = content.replace("uses: pnpm/action-setup@v2\n        with:\n          version: '9.12.0'\n", "uses: pnpm/action-setup@v2\n")
    content = content.replace("uses: pnpm/action-setup@v4\n        with:\n          version: \"9.12.0\"\n", "uses: pnpm/action-setup@v4\n")
    content = content.replace("uses: pnpm/action-setup@v3\n        with:\n          version: \"9.12.0\"\n", "uses: pnpm/action-setup@v3\n")
    content = content.replace("uses: pnpm/action-setup@v2\n        with:\n          version: \"9.12.0\"\n", "uses: pnpm/action-setup@v2\n")

    content = content.replace("uses: pnpm/action-setup@v4\n        with: { version: 9.12.0 }\n", "uses: pnpm/action-setup@v4\n")
    content = content.replace("uses: pnpm/action-setup@v3\n        with: { version: 9.12.0 }\n", "uses: pnpm/action-setup@v3\n")

    content = content.replace("PNPM_VERSION: '9.12.0'", 'PNPM_VERSION: "10.0.0"')
    content = content.replace('PNPM_VERSION: "9.12.0"', 'PNPM_VERSION: "10.0.0"')
    content = content.replace('pnpm-version:\n        required: false\n        type: string\n        default: \'9.12.0\'', 'pnpm-version:\n        required: false\n        type: string\n        default: \'10.0.0\'')
    content = content.replace('pnpm-version:\n        description: "Version of pnpm to use"\n        required: false\n        type: string\n        default: "9.12.0"', 'pnpm-version:\n        description: "Version of pnpm to use"\n        required: false\n        type: string\n        default: "10.0.0"')
    content = content.replace('pnpm_version:\n        description: "Version of pnpm to use"\n        required: false\n        type: string\n        default: "9.12.0"', 'pnpm_version:\n        description: "Version of pnpm to use"\n        required: false\n        type: string\n        default: "10.0.0"')

    with open(wf, 'w') as f:
        f.write(content)

File: fix_pnpm.mjs
import fs from 'node:fs';
import path from 'node:path';

function walk(dir) {
  let results = [];
  const list = fs.readdirSync(dir);
  list.forEach((file) => {
    file = dir + '/' + file;
    const stat = fs.statSync(file);
    if (stat && stat.isDirectory()) {
      results = results.concat(walk(file));
    } else {
      if (file.endsWith('.yml') || file.endsWith('.yaml')) {
        results.push(file);
      }
    }
  });
  return results;
}

const files = walk('.github/workflows');
let modifiedFiles = 0;

for (const file of files) {
  let content = fs.readFileSync(file, 'utf8');
  let lines = content.split('\n');
  let newLines = [];
  let pnpmSetupIndex = -1;
  let pnpmSetupLines = [];
  let nodeSetupIndex = -1;
  let nodeSetupLines = [];

  // A more robust way is to just do a simple search and replace or match blocks,
  // but looking at the error, it's specific to the 'Quarantine Tests (Flaky)'
  // and 'Security Gate (Gitleaks + Snyk)' jobs.
  // Wait, looking at the previous grep, the order is actually:
  // pnpm/action-setup
  // actions/setup-node
  // So the problem might be that setup-node is not finding pnpm because we didn't specify the `cache: 'pnpm'` correctly,
  // or because of `version` in pnpm/action-setup being removed so it defaults to packageManager from package.json,
  // but node is not set up before pnpm/action-setup?
  // Let me read the repository memory carefully:
  // "For GitHub Actions workflows, always use pnpm/action-setup combined with pnpm install.
  // It must run *before* actions/setup-node to avoid 'Unable to locate executable file: pnpm' errors.
  // Do not hardcode the version in pnpm/action-setup if packageManager is defined in package.json."
}
